From a71659231c9cb6d35b1a8ad20f553096df2c56f8 Mon Sep 17 00:00:00 2001
From: tbordaz <tbordaz@redhat.com>
Date: Mon, 3 Nov 2025 11:57:15 +0100
Subject: [PATCH] Issue 7065 - A search filter containing a non normalized DN
 assertion does not return matching entries (#7068)

Bug description:
	This bug is a regression introduced with #6172.
	Since the filter was optimized #5170, in case of filter bypass
	the client filter remains not normalized. For access control purpose
	it is nevertheless checked.
	Checking main fail if the normalized sorted valueset does not match
	unnormalized client assertion

Fix description:
	Normalize the filter and intended filter before using it for
	matching

fixes: #7065

Reviewed by: Pierre Rogier, Mark Reynolds (thanks)
---
 .../tests/suites/filter/basic_filter_test.py  | 33 +++++++++++++++++++
 ldap/servers/slapd/back-ldbm/ldbm_search.c    |  4 +++
 2 files changed, 37 insertions(+)

diff --git a/dirsrvtests/tests/suites/filter/basic_filter_test.py b/dirsrvtests/tests/suites/filter/basic_filter_test.py
index b758cf884..92af6f9d8 100644
--- a/dirsrvtests/tests/suites/filter/basic_filter_test.py
+++ b/dirsrvtests/tests/suites/filter/basic_filter_test.py
@@ -6,12 +6,14 @@
 # See LICENSE for details.
 # --- END COPYRIGHT BLOCK ----
 
+import ldap
 import pytest, os
 
 from lib389._constants import DEFAULT_SUFFIX, PW_DM
 from lib389.topologies import topology_st as topo
 
 from lib389.idm.user import UserAccount, UserAccounts
+from lib389.idm.group import UniqueGroups
 from lib389.idm.account import Accounts
 
 pytestmark = pytest.mark.tier0
@@ -43,6 +45,37 @@ def test_search_attr(topo):
     # Testing filter is working for other filters
     assert len(user.filter("(objectclass=inetOrgPerson)")) == 4
 
+def test_filter_assertion_not_normalized(topo, request):
+    """Test ldapsearch with a filter with an assertion value
+    that is not normalize
+
+    :id: c23366aa-d134-4f95-86ee-44ad3e05480b
+    :setup: Standalone instance
+    :steps:
+         1. Adding 20 member values to a group
+         2. Checking the a search with not normalize
+            assertion returns the group
+    :expectedresults:
+         1. This should pass
+         2. This should pass
+    """
+    groups = UniqueGroups(topo.standalone, DEFAULT_SUFFIX)
+    demo_group = groups.ensure_state(properties={'cn': 'demo_group'})
+
+    # Adding more that 10 values to 'member' so that
+    # the valueset will be sorted
+    values = []
+    for idx in range(0,20):
+        values.append('uid=scarter%d, ou=People, dc=example,dc=com' % idx)
+    demo_group.add('member', values)
+
+    entries = topo.standalone.search_s(DEFAULT_SUFFIX, ldap.SCOPE_SUBTREE, 'member=uid=scarter0,   ou=People,dc=example,dc=com')
+    assert len(entries) == 1
+
+    def fin():
+        demo_group.delete('member')
+
+    request.addfinalizer(fin)
 
 if __name__ == "__main__":
     CURRENT_FILE = os.path.realpath(__file__)
diff --git a/ldap/servers/slapd/back-ldbm/ldbm_search.c b/ldap/servers/slapd/back-ldbm/ldbm_search.c
index 2add5ec38..c18eac40b 100644
--- a/ldap/servers/slapd/back-ldbm/ldbm_search.c
+++ b/ldap/servers/slapd/back-ldbm/ldbm_search.c
@@ -1083,6 +1083,8 @@ build_candidate_list(Slapi_PBlock *pb, backend *be, struct backentry *e, const c
         slapi_filter_optimise(filter);
         /* modify the filter to be: (&(|(originalfilter)(objectclass=referral))(parentid=idofbase)) */
         filter_exec = create_onelevel_filter(filter, e, managedsait);
+        slapi_filter_normalize(filter, PR_TRUE /* normalize values too */);
+        slapi_filter_normalize(filter_exec, PR_TRUE /* normalize values too */);
 
         slapi_log_err(SLAPI_LOG_FILTER, "ldbm_back_search", "Optimised ONE filter to - %s\n",
              slapi_filter_to_string(filter_exec, logbuf, sizeof(logbuf)));
@@ -1097,6 +1099,7 @@ build_candidate_list(Slapi_PBlock *pb, backend *be, struct backentry *e, const c
     case LDAP_SCOPE_SUBTREE:
         /* Now optimise the filter for use */
         slapi_filter_optimise(filter);
+        slapi_filter_normalize(filter, PR_TRUE /* normalize values too */);
 
         slapi_pblock_get(pb, SLAPI_OPERATION, &operation);
         if (!slapi_be_is_flag_set(be, SLAPI_BE_FLAG_CONTAINS_REFERRAL) || (operation && operation_is_flag_set(operation, OP_FLAG_INTERNAL))) {
@@ -1108,6 +1111,7 @@ build_candidate_list(Slapi_PBlock *pb, backend *be, struct backentry *e, const c
         } else {
             /* make (|(originalfilter)(objectclass=referral)) */
             filter_exec = create_subtree_filter(filter, managedsait);
+            slapi_filter_normalize(filter_exec, PR_TRUE /* normalize values too */);
         }
 
         slapi_log_err(SLAPI_LOG_FILTER, "ldbm_back_search", "Optimised SUB filter to - %s\n",
-- 
2.52.0

